src/Microsoft.Data.Analysis/DataFrame.cs

```diff
@@ -316,14 +316,28 @@ public DataFrame AddSuffix(string suffix, bool inPlace = false)
         /// <param name="numberOfRows">Number of rows in the returned DataFrame</param>
         public DataFrame Sample(int numberOfRows)
         {
+            if (numberOfRows > Rows.Count)
+            {
+                throw new ArgumentException(string.Format(Strings.ExceedsNumberOfRows, Rows.Count), nameof(numberOfRows));
+            }
+
+            int shuffleLowerLimit = 0;
+            int shuffleUpperLimit = (int)Math.Min(Int32.MaxValue, Rows.Count);
+            
+            int[] shuffleArray = Enumerable.Range(0, shuffleUpperLimit).ToArray();
             Random rand = new Random();
-            PrimitiveDataFrameColumn<long> indices = new PrimitiveDataFrameColumn<long>("Indices", numberOfRows);
-            int randMaxValue = (int)Math.Min(Int32.MaxValue, Rows.Count);
-            for (long i = 0; i < numberOfRows; i++)
+            while (shuffleLowerLimit < numberOfRows)
             {
-                indices[i] = rand.Next(randMaxValue);
+                int randomIndex = rand.Next(shuffleLowerLimit, shuffleUpperLimit);
+                int temp = shuffleArray[shuffleLowerLimit];
+                shuffleArray[shuffleLowerLimit] = shuffleArray[randomIndex];
+                shuffleArray[randomIndex] = temp;
+                shuffleLowerLimit++;
             }
+            ArraySegment<int> segment = new ArraySegment<int>(shuffleArray, 0, shuffleLowerLimit);
 
+            PrimitiveDataFrameColumn<int> indices = new PrimitiveDataFrameColumn<int>("indices", segment);
```

Comment:
Ah, ideally, we'd have had a `DataFrame.Clone` overload that takes an `IEnumerable<int>` already so we didn't need to allocate another indices column here. If you're interested, we can create another issue for it?

Reply:
Will raise the issue now. Thanks for the help and review!